Add option for eBPF plugin to manipulate memory

thiagoftsm opened the pull request:
##### Summary
This PR is bringing another method to adjust the memory usage for eBPF.plugin. This new method will be set by default, because it needs less iteration with the kernel, on the other hand it will use more memory from computer. My initial conclusion is that it worth while to use more memory and speed up process and environment. We are also adding a new parameter that allows user to control the memory mode.

![fix](https://user-images.githubusercontent.com/49162938/165199780-4c9476b7-644c-492d-b688-77c2d39ed8be.png)


##### Test Plan

1. Update your `/etc/netdata/ebpf.d.conf` adding the new parameter:
```conf
[global]
    allocate memory = static
```
2. Clone this branch and compile it.
3. Run Netdata and you will see that eBPF.plugin will have memory and cpu usage smaller than `apps`. Depending of your kernel version, the same problems reported [here](https://github.com/netdata/netdata/issues/12719#issuecomment-1112808123) will happen.
4. Run `curl -k -o FILE.txt "https://localhost:19999/api/v1/data?chart=apps.mem&dimension=ebpf.plugin"` after few minutes, and you will see a "stable" `memory`.
5. Stop netdata, and set adjust your `/etc/netdata/ebpf.d.conf`:
```conf
[global]
    allocate memory = dynamic
```
6. Run step 4 again, and values won't be so stable.

We expect to have data collection normally with both options. Of course with `dynamic` we expect major iteration with kernel and more CPU and memory usage.

##### Additional Information
It is not necessary to test on different kernels, because we are changing only `user ring` code.

The following table demonstrate the results using current `master` branch and this PR:

| Branch | Result |
|------------|----------|
| `master` | [old.txt](https://github.com/netdata/netdata/files/8602063/old.txt) |
| This PR |  [new.txt](https://github.com/netdata/netdata/files/8602065/new.txt)|


<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ebpf.plugin
- Can they see the change or is it an under the hood? If they can see it, where? Users will have less memory and cpu usage for `ebpf.plugin`
- How is the user impacted by the change?  His machine will have more resources for other tasks.
- What are there any benefits of the change?  we are improving data collection
</details>

vlvkobal commented:
Do you have an idea of what's going on here?
```
libbpf: failed to find kernel BTF type ID of 'netdata_sync': -3
libbpf: prog 'netdata_sync_fentry': failed to prepare load attributes: -3
libbpf: failed to load program 'netdata_sync_fentry'
libbpf: failed to load object 'sync_bpf'
libbpf: failed to load BPF skeleton 'sync_bpf': -3
```
```
==24746== Conditional jump or move depends on uninitialised value(s)
==24746==    at 0x484CBDC: strcmp (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==24746==    by 0x1745E5: load_module_btfs (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x17E17F: find_kernel_btf_id (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x184AC2: bpf_program__set_attach_target (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x12E2C9: ebpf_sync_load_and_attach (ebpf_sync.c:160)
==24746==    by 0x12E2C9: ebpf_sync_initialize_syscall (ebpf_sync.c:249)
==24746==    by 0x12E64A: ebpf_sync_thread (ebpf_sync.c:572)
==24746==    by 0x1542A8: thread_start (threads.c:185)
==24746==    by 0x4E5C5C1: start_thread (in /usr/lib/libc.so.6)
==24746==    by 0x4EE1583: clone (in /usr/lib/libc.so.6)
```
```
==24746== Use of uninitialised value of size 8
==24746==    at 0x15D53A: btf_parse_str_sec (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x15E9F0: btf_new (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x15FC2F: btf_get_from_fd (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x174619: load_module_btfs (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x17E17F: find_kernel_btf_id (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x184AC2: bpf_program__set_attach_target (in /home/vlad/netdata/ebpf.plugin)
==24746==    by 0x12E2C9: ebpf_sync_load_and_attach (ebpf_sync.c:160)
==24746==    by 0x12E2C9: ebpf_sync_initialize_syscall (ebpf_sync.c:249)
==24746==    by 0x12E64A: ebpf_sync_thread (ebpf_sync.c:572)
==24746==    by 0x1542A8: thread_start (threads.c:185)
==24746==    by 0x4E5C5C1: start_thread (in /usr/lib/libc.so.6)
==24746==    by 0x4EE1583: clone (in /usr/lib/libc.so.6)
```

thiagoftsm commented:
> Do you have an idea of what's going on here?

Firstly, thank you to report the warnings.

I took a look in the code and I created this PR https://github.com/netdata/netdata/pull/12831 to fix the issue.

I could not recreate the problems reported by `valgrind`, but there was only one variable that was not initialized in the beginning of the thread, but the variable was filled before the function to be called.

Finally all the warnings reported do not have relationship with this PR.

Best regards!

vlvkobal commented on collectors/ebpf.plugin/README.md:
```diff
@@ -203,6 +203,37 @@ Linux metrics:
+| swap      |           16 | 262144 |   786432|
+| vfs       |          112 | 262144 |  3932160|
+
+Dynamic allocates the minimum necessary by default, this value is calculated multiplying the total number of PIDs (`32768`)
+times 8 bytes (64 bits).
+
+Static will allocate everything necessary to process when the collector starts. The value for this column is
+calculated adding the value of `Static` to the result of multiplying the column `Size (bytes)` and the number of PIDs (`32768`).
```
I'm not sure this part is needed at all.

vlvkobal replied:
I really doubt the PR fixes the bug, discussed in #12719. @vobruba-martin, could you test the PR? If it doesn't fix the bug, I'm not sure it brings much value. @vkalintiris, what do you think?

thiagoftsm replied:
I sent this information for documentation team and we decided to give the maximum information possible for users to understand how we calculated the values.

thiagoftsm replied:
@DShreve2 what do you think about @vlvkobal suggestion?

vlvkobal commented:
I really doubt the PR fixes the bug, discussed in #12719. @vobruba-martin, could you test the PR? If it doesn't fix the bug, I'm not sure it brings much value. @vkalintiris, what do you think?

thiagoftsm commented:
Hello @vlvkobal ,

Let's go by steps. I force a rebase the PR to remove the `sync` warnings, and I already addressed all your suggestions for the documentation.

> I really doubt the PR fixes the bug, discussed in #12719. @vobruba-martin, could you test the PR? If it doesn't fix the bug, I'm not sure it brings much value. @vkalintiris, what do you think?

As I already answered @vobruba-martin , and you can see [here](https://github.com/netdata/netdata/issues/12719#issuecomment-1112808123), this PR is reducing interaction with kernel and as consequence we are reducing CPU usage and we are not more allocating memory for every new PID. Since yesterday I started working with the last step that will fix definitively not only the issue #12719, but it will speed up the shutdown process. I did not bring everything in this PR, because it is already too big, and I have work to do in two other repository before to create the last PR.

Best regards!

ilyam8 commented on collectors/ebpf.plugin/README.md:
```diff
@@ -203,6 +203,37 @@ Linux metrics:
+The following table demonstrates a comparison between memory usage for each method on a 64 bit computer running a Linux 
+distribution that allows the maximum of `32768` PIDs (Value obtained from `/proc/sys/kernel/pid_max`). 
+
+| Thread  | Strcuture size | Dynamic preallocation | Static allocation |
+|-----------|--------------|--------|---------|
```
```suggestion
| Thread  | Structure size (B) | Dynamic preallocation (MB) | Static allocation (MB) |
|-----------|--------------|--------|---------|
```

(And let's converts MB values to megabytes)

thiagoftsm replied:
Thanks!

vobruba-martin commented:
I did a quick test and it doesn't seem to me that there is any difference. I will let it run for a day or two to see if the increased CPU usage is static or not.

<details>
<summary>CPU usage charts</summary>
![image](https://user-images.githubusercontent.com/6897988/167289896-c43f8834-b73f-4030-b4d3-bb60e8c162ca.png)
![image](https://user-images.githubusercontent.com/6897988/167289911-3fa5b42c-fe3f-44c7-9064-36f6e7912241.png)
</details>

Or did I something wrong?
1. I followed [this guide](https://learn.netdata.cloud/docs/agent/packaging/installer/methods/manual#install-netdata) to install Netdata from Git.
2. Before running `./netdata-installer.sh`:
    1. `git remote add thiagoftsm https://github.com/thiagoftsm/netdata.git`
    2. `git fetch thiagoftsm`
    3. `git merge thiagoftsm/dynamic_or_static`

vobruba-martin commented:
Sadly, today I see no difference compared to stable version. Look what is does if I just restart Netdata.

<details>
<summary>CPU charts</summary>

![image](https://user-images.githubusercontent.com/6897988/167366219-2bd54eae-08f7-4561-9403-670e53d0a19d.png)
</details>

thiagoftsm commented:
Hello @vobruba-martin ,

> Sadly, today I see no difference compared to stable version. Look what is does if I just restart Netdata.

The main goal of this PR is to reduce memory allocation and keep it static on time, please, can you take a look in the memory to confirm this was fixed?

I updated the OP removing the "Fixes", because we will finally fix with next PR.

Best regards!

vobruba-martin commented:
Hello @thiagoftsm ,

memory usage doesn't seem to be constant in time.

<details>
<summary>Memory chart</summary>

![image](https://user-images.githubusercontent.com/6897988/167563106-fe13ac81-66d5-4445-9ab3-d529ce09b9bd.png)
</details>

ilyam8 commented:
@vobruba-martin any chance you can set up a lab with VM with ~workload so the bug shows up and give @thiagoftsm access to it? It would be great and much easier to pinpoint the problem.

vobruba-martin commented:
@ilyam8 Good idea! I will give it a try if I have some time in upcoming days.

thiagoftsm commented:
Hello guys,

We are already merging the solution to reduce process time, the first PR (https://github.com/netdata/kernel-collector/pull/302) was already merged today.  Next days I will define if I will bring the final update for this PR or if I will  keep my initial conclusion that we need another PR that will fix problems definitively. 

Best regards!

thiagoftsm commented:
Closing in favor of https://github.com/netdata/netdata/pull/13397. 
Some of memory leaks were fixed with https://github.com/netdata/netdata/pull/13407.
